infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/transferprocessor/ParallelTransferProcessor.java

```diff
@@ -0,0 +1,139 @@
+import static org.corfudb.infrastructure.log.statetransfer.transferprocessor.TransferProcessor.TransferProcessorResult.TransferProcessorStatus.TRANSFER_FAILED;
+import static org.corfudb.infrastructure.log.statetransfer.transferprocessor.TransferProcessor.TransferProcessorResult.TransferProcessorStatus.TRANSFER_SUCCEEDED;
+
+/**
+ * A transfer processor that performs state transfer by dynamically distributing and parallelizing
+ * the workload among the source log unit servers of each segment.
+ */
+@Slf4j
+public class ParallelTransferProcessor implements TransferProcessor {
+    /**
+     * A state transfer batch processor that performs the batch transfer.
+     */
+    private final StateTransferBatchProcessor stateTransferBatchProcessor;
+
+    /**
+     * A number of in-flight requests per one node.
+     */
+    private static final int NUM_REQUESTS_PER_NODE = 5;
+
+    public ParallelTransferProcessor(StateTransferBatchProcessor stateTransferBatchProcessor) {
+        this.stateTransferBatchProcessor = stateTransferBatchProcessor;
+    }
+
+    private CompletableFuture<Void> handleNonTransferException(CompletableFuture<Void> futures,
+                                                               Throwable ex) {
+        CompletableFuture<Void> failedFuture = new CompletableFuture<>();
+        failedFuture.completeExceptionally(ex);
+        return CFUtils.allOfOrTerminateExceptionally(futures, failedFuture);
+    }
+
+    private CompletableFuture<Void> handleBatchRequest(TransferBatchRequest request,
+                                                       CompletableFuture<Void> allFutures,
+                                                       Semaphore semaphore) throws InterruptedException {
+        // If the request is of type SEGMENT_INIT - it signals to the batch processor that the
+        // consecutive DATA requests all belong to the same segment. Since all the DATA requests
+        // that belong to the same segment have the same number of source nodes, we can dynamically
+        // configure the semaphore to support the average parallelism level for the entire segment
+        // which is equal to the NUM_REQUEST_PER_NODE * number of source nodes in the segment.
+        if (request.getBatchType() == SEGMENT_INIT) {
+            int numPermits = request.getDestinationNodes()
+                    .map(AbstractCollection::size).orElse(1) *
+                    NUM_REQUESTS_PER_NODE;
+            if (numPermits == 0) {
+                throw new IllegalStateException("Number of permits should be equal to" +
+                        " at least one.");
+            }
+            log.trace("Starting a new segment transfer. Parallelism level: {}", numPermits);
+            // Acquire all the permits if any.
+            semaphore.drainPermits();
+            // Release numPermits for the batch transfer futures to acquire.
+            semaphore.release(numPermits);
```

Comment:
This is quite hacky and probably incorrect. drainPermits() only drains currently "available" permits, after that previously acquired permits can still be released, reusulting in more permits than numPermits here. I think we need to transfer segments one by one and within each segment, you can parallelize it with a fixed number of permits.

Reply:
@WenbinZhu Ok, it makes sense. I split the workload by segments. Please see if the code makes sense.